HeterogeneousCore/AlpakaInterface/interface/workdivision.h

```diff
@@ -496,6 +507,219 @@ namespace cms::alpakatools {
     const Idx range_;
   };
 
+  /* uniform_groups
+   *
+   * `uniform_groups(acc, elements)` returns a range than spans the group indices required to cover the given problem
+   * size, in units of the block size:
+   *   - the `elements` argument indicates the total number of elements, across all groups.
+   *
+   * `uniform_groups` should be called consistently by all the threads in a block. All threads in a block see the same
+   * loop iterations, while threads in different blocks may see a different number of iterations.
+   *
+   * For example, if `size` is 1000 and the block size is 16,
+   *
+   *   for (auto group: uniform_groups(acc, 1000)
+   *
+   * will return the range from 0 to 62, split across all blocks in the work division.
+   *
+   * If the work division has more than 63 blocks, the first 63 will perform one iteration of the loop, while the other
+   * blocks will exit immediately.
+   * If the work division has less than 63 blocks, some of the blocks will perform more than one iteration, in order to
+   * cover then whole problem space.
```

Comment:
No, it's as the comment says: _all threads_ (0-15) will execute the loop body.

It's up to the inner loop (or the user, if they don't use an inner loop) to run only on a subset of the threads.

In fact,
```c++
for (auto group: uniform_groups(acc, 1000) { ... }
```
should be exactly the same as
```c++
for (auto group: independent_groups(acc, round_up_by(1000, alpaka::getWorkDiv<alpaka::Grid, alpaka::Elems>)) { ... }
```

Reply:
Ah right, this loop is about the groups/blocks, and not about elements, thanks.

So if one would do
```cpp
for (auto group : uniform_groups(acc, 1000) {
  for (auto element : uniform_group_elements(acc, group, 1000) {
     ...
  }
}
```
it's the inner loop (`uniform_group_elements`) where one needs to be careful with `alpaka::syncBlockThreads()`, right?

And there it would happen that the threads 0-7 of the block executing group 62 would run the inner loop, and there threads 8-15 would not, right?

 If this is the case, how about adding a warning in the comment of `uniform_group_elements` (perhaps also for `elements_in_block`) that the `alpaka::syncBlockThreads()` is not generally safe to be called within that loop?

Reply:
> it's the inner loop (`uniform_group_elements`) where one needs to be careful with `alpaka::syncBlockThreads()`, right?

Yes.

In fact, it does not make sense to call `alpaka::syncBlockThreads()` within the inner loop.
One should always split the inner loop in multiple loops, and synchronise between them:
```c++
for (auto group : uniform_groups(acc, 1000) {
  for (auto element : uniform_group_elements(acc, group, 1000) {
     // no synchronisations here
     ...
  }
  alpaka::syncBlockThreads();
  for (auto element : uniform_group_elements(acc, group, 1000) {
     // no synchronisations here
     ...
  }
  alpaka::syncBlockThreads();
}
```

Reply:
> And there it would happen that the threads 0-7 of the block executing group 62 would run the inner loop, and there threads 8-15 would not, right?

All threads would execute the `for (auto element : uniform_group_elements(acc, group, 1000)` loop.
However, the number of iterations of the inner loop will be different for the different threads: 0-7 will execute 1 iteration, 8-15 will execute 0 iterations.

Reply:
> If this is the case, how about adding a warning in the comment of `uniform_group_elements` (perhaps also for `elements_in_block`) that the `alpaka::syncBlockThreads()` is not generally safe to be called within that loop?

Sure, see https://github.com/cms-sw/cmssw/pull/43632 .

Reply:
> However, the number of iterations of the inner loop will be different for the different threads: 0-7 will execute 1 iteration, 8-15 will execute 0 iterations.

Thanks, this is what I was after, but written more accurately.